File: cpp/series/series.cpp
#include "series.hpp"

namespace series {

std::vector<std::string> slice(std::string sequence, int span) {
    std::vector<std::string> groups{};

    if (span == 0) {
        throw std::domain_error{"span can't be zero"};
    }

    if (span < 0) {
        throw std::domain_error{"span can't be negative"};
    }

    if (sequence.empty()) {
        throw std::domain_error{"sequence can't be empty"};
    }

    if ((long long)sequence.length() < span) {
        throw std::domain_error{"span is longer than sequence"};
    }

    if ((long long)sequence.length() == span) {
        groups.emplace_back(sequence);
        return groups;
    }

    size_t remaining = sequence.length();
    auto current = sequence.begin();

    for (const char digit : sequence) {
        if (remaining < (size_t)span) {
            break;
        }
        remaining -= 1;

        auto next = current;
        std::string group{};

        group += digit;

        for (int i = 1; i < span; i++) {
            std::advance(next, 1);
            group += *next;
        }

        groups.emplace_back(group);

        std::advance(current, 1);
    }

    return groups;
}

} // namespace series

File: ruby/series/series.rb
# frozen_string_literal: false

# https://exercism.org/tracks/ruby/exercises/series
# Series exercise
class Series
  def initialize(sequence)
    @sequence = sequence
    @sequence_length = @sequence.length
    @span = 0
  end

  def slices(span)
    @span = span

    raise ArgumentError, "span can't be zero" if @span.zero?
    raise ArgumentError, "span can't be negative" if @span.negative?
    raise ArgumentError, "sequence can't be empty" if @sequence.empty?
    raise ArgumentError, 'span is longer than sequence' if @span > @sequence_length

    groups = []

    return groups.push(@sequence) if @span == @sequence_length

    remaining = @sequence_length

    (0...@sequence_length).each do |index|
      break if remaining < span

      remaining -= 1

      groups.push(@sequence.slice(index, @span))
    end

    groups
  end
end
